### PR TITLE
enhancement: enable icap preview mode and fix client according to the spec

### PR DESCRIPTION
**Pull request by fschade**
## Description
the used icap lib is not active maintained and we've spotted some bugs:
* TCP socket keepalive
* broken preview requests

both of the above topics are handled in my fork but needs confirmation and polishing before sending the changes upstream.

The implementation for preview requests follows the [rfc 3507](https://datatracker.ietf.org/doc/html/rfc3507) (0 DoubleCRLF) and works (at least in my test-cases) now as expected.

Making use of the preview mechanism could drastically speed up virus detection and therefore improves the end user experience and by reducing the waiting time and on also reduces the network traffic (if a virus is detected within the firs N bytes).

The implementation still needs some more options later on, for now i just wanna wait for confirmation that the bug reported via https://github.com/owncloud/ocis/issues/6764 is fixed, Once that is done, i start polishing.

The preview requests are enabled by default in this PR, no change needed.
To see the benefit of TCP connection keepalive, changes in the c-icap are needed, e.g.:  `KeepAliveTimeout 600`

## Related Issue
- Fixes https://github.com/owncloud/ocis/issues/6764

## Motivation and Context
speedup virus scanning and improve end user experience

## How Has This Been Tested?
- locally only, needs confirmation by devops team 

## Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: services/antivirus/pkg/service/service.go [hunk 1]
```diff
@@ -16,34 +16,44 @@ import (
...
+	var scanner Scanner
 	var err error
-	av.s, err = scanners.New(c)
+	switch c.Scanner.Type {
+	default:
+		return Antivirus{}, fmt.Errorf("unknown av scanner: '%s'", c.Scanner.Type)
+	case "clamav":
+		scanner = scanners.NewClamAV(c.Scanner.ClamAV.Socket)
+	case "icap":
+		scanner, err = scanners.NewICAP(c.Scanner.ICAP.URL, c.Scanner.ICAP.Service, time.Duration(c.Scanner.ICAP.Timeout)*time.Second)
```
Comment (kobergj) [inline services/antivirus/pkg/service/service.go:49]: We should change config value to be of type `time.Duration`. (Needs deprecation maybe.) That way users could define the timeout like `30s`. What do you think?

**Review by kobergj — APPROVED** [review]
:+1: LGTM 

Just some remark for future improvement...

File: services/antivirus/pkg/service/service.go (hunk 1 above)
Comment (fschade) [reply services/antivirus/pkg/service/service.go:49]: yup, needs further polishing... happy to take care once we see it does what it should

**Comment by fschade** [comment]
i updated the client lib, can be found here: https://github.com/fschade/icap-client/pull/1/

File: services/antivirus/pkg/config/config.go [hunk 2]
```diff
@@ -70,9 +71,9 @@ type ClamAV struct {
 	Socket string `yaml:"socket" env:"ANTIVIRUS_CLAMAV_SOCKET" desc:"The socket clamav is running on. Note the default value is an example which needs adaption according your OS."`
 }
 
-// ICAP provides configuration option for ICAP
+// ICAP provides configuration options for icap
 type ICAP struct {
-	Timeout int64  `yaml:"timeout" env:"ANTIVIRUS_ICAP_TIMEOUT" desc:"Timeout for the ICAP client."`
-	URL     string `yaml:"url" env:"ANTIVIRUS_ICAP_URL" desc:"URL of the ICAP server."`
-	Service string `yaml:"service" env:"ANTIVIRUS_ICAP_SERVICE" desc:"The name of the ICAP service."`
+	Timeout time.Duration `yaml:"timeout" env:"ANTIVIRUS_ICAP_TIMEOUT" desc:"Timeout for the ICAP client. Defaults to '5m' (5 minutes)"`
```
Comment (butonic) [inline services/antivirus/pkg/config/config.go:76]: is this a breaking change? we certainly now allow time dimensions, but how will this interpret existing configs?

**Review by butonic — APPROVED** [review]

File: services/antivirus/pkg/config/config.go (hunk 2 above)
Comment (fschade) [reply services/antivirus/pkg/config/config.go]: yes, good catch, i deprecated it as described here:
https://owncloud.dev/ocis/development/deprecating-variables/

beside the deprecation it also contains a migration step, so not breaking anymore

**Comment by fschade** [comment]
@mmattel, please see the message from jfd regarding the deprecation.
It contains a new env var for the timeout, e.g. `ANTIVIRUS_ICAP_SCAN_TIMEOUT`

**Comment by fschade** [comment]
the client lib is in a good shape now, still we could (and should) invest more time in integrate a protocol layer or similar.
Can be done later!

we have to decide if we backport it or not, my personal opinion, 4.0, master and 4.0 should contain the change.

i tested (locally) every known case, works for me! cc.: @wkloucek. please confirm if it solves your icap problems.

File: services/antivirus/pkg/config/config.go [hunk 3]
```diff
@@ -70,9 +71,10 @@ type ClamAV struct {
...
 }
 
-// ICAP provides configuration option for ICAP
+// ICAP provides configuration options for icap
 type ICAP struct {
-	Timeout int64  `yaml:"timeout" env:"ANTIVIRUS_ICAP_TIMEOUT" desc:"Timeout for the ICAP client."`
-	URL     string `yaml:"url" env:"ANTIVIRUS_ICAP_URL" desc:"URL of the ICAP server."`
-	Service string `yaml:"service" env:"ANTIVIRUS_ICAP_SERVICE" desc:"The name of the ICAP service."`
+	DeprecatedTimeout int64         `yaml:"timeout" env:"ANTIVIRUS_ICAP_TIMEOUT" desc:"Timeout for the ICAP client." deprecationVersion:"5.0" removalVersion:"6.0" deprecationInfo:"changing the type for consistency" deprecationReplacement:"ANTIVIRUS_ICAP_SCAN_TIMEOUT"`
+	Timeout           time.Duration `yaml:"scan_timeout" env:"ANTIVIRUS_ICAP_SCAN_TIMEOUT" desc:"Timeout for the ICAP client. Defaults to '5m' (5 minutes)"`
```
Comment (mmattel) [inline services/antivirus/pkg/config/config.go:77]:
```suggestion
	DeprecatedTimeout int64         `yaml:"timeout" env:"ANTIVIRUS_ICAP_TIMEOUT" desc:"Timeout for the ICAP client." deprecationVersion:"5.0" removalVersion:"6.0" deprecationInfo:"Changing the envvar type for consistency reasons." deprecationReplacement:"ANTIVIRUS_ICAP_SCAN_TIMEOUT"`
	Timeout           time.Duration `yaml:"scan_timeout" env:"ANTIVIRUS_ICAP_SCAN_TIMEOUT" desc:"Scan timeout for the ICAP client. Defaults to '5m' (5 minutes). See the Environment Variable Types description for more details."`
```

**Comment by mmattel** [comment]
Adding @ScharfViktor as this is testing relevant
We also need to add a upgrade not into the upgrade section of the admin docs.

**Comment by fschade** [comment]
thanks martin for taking care :)

**Comment by mmattel** [comment]
@fschade we need a changelog!

File: services/antivirus/pkg/config/config.go (hunk 3 above)
Comment (mmattel) [inline services/antivirus/pkg/config/config.go:77]:
```suggestion
	DeprecatedTimeout int64         `yaml:"timeout" env:"ANTIVIRUS_ICAP_TIMEOUT" desc:"Timeout for the ICAP client." deprecationVersion:"5.0" removalVersion:"6.0" deprecationInfo:"Changing the envvar type for consistency reasons." deprecationReplacement:"ANTIVIRUS_ICAP_SCAN_TIMEOUT"`
	Timeout           time.Duration `yaml:"scan_timeout" env:"ANTIVIRUS_ICAP_SCAN_TIMEOUT" desc:"Scan timeout for the ICAP client. Defaults to '5m' (5 minutes). See the Environment Variable Types description for more details."`
```
By adding the changelog without rebasing, the text changes commited got reset - adding again.

**Review by mmattel — APPROVED** [review]
LGTM from a docs pov 👍

File: services/antivirus/pkg/config/parser/parse.go [hunk 4]
```diff
@@ -34,5 +36,10 @@ func ParseConfig(cfg *config.Config) error {
 
 // Validate validates our little config
 func Validate(cfg *config.Config) error {
+	if cfg.Scanner.ICAP.DeprecatedTimeout != 0 {
+		cfg.Scanner.ICAP.Timeout = time.Duration(cfg.Scanner.ICAP.DeprecatedTimeout) * time.Second
+		fmt.Println("ANTIVIRUS_ICAP_TIMEOUT is deprecated, use ANTIVIRUS_ICAP_SCAN_TIMEOUT instead")
```
Comment (butonic) [inline services/antivirus/pkg/config/parser/parse.go:41]: I see what you did there ❤️

**Pull request merged**